File: src/graalvm/gmicronaut/src/main/java/com/benchmark/service/DatabaseService.java
package com.benchmark.service;

import com.benchmark.model.User;
import com.benchmark.model.UserStats;
import io.micronaut.context.annotation.Primary;
import jakarta.inject.Inject;
import jakarta.inject.Named;
import jakarta.inject.Singleton;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Reads {@code users}/{@code orders} over raw JDBC.
 *
 * <p>These two queries used to live in a Micronaut Data
 * {@code @JdbcRepository} ({@code UserRepository}). The
 * {@code micronaut-data-processor} implements a repository interface at
 * compile time by introspecting the entity, which requires the return type to
 * carry {@code @MappedEntity} (+ {@code @Id}). {@code User}/{@code UserStats}
 * are plain Jackson serialization DTOs with no JPA-style mapping on purpose
 * (every other implementation here maps rows by hand), so the processor aborted
 * the build with "Could not resolve root entity" for {@code findByIdRaw}.
 * Rather than pollute the DTOs with Micronaut Data annotations, the queries are
 * executed here against the {@code default} {@link DataSource} produced by
 * {@link com.benchmark.config.DatasourceFactory}. The SQL is byte-for-byte the
 * normative query from {@code contracts/rest/canonical-payloads.md}.
 */
@Singleton
@Primary
public class DatabaseService {
    private final DataSource dataSource;

    @Inject
    public DatabaseService(@Named("default") DataSource dataSource) {
        this.dataSource = dataSource;
    }

    public Optional<User> getUserById(Integer id) {
        final String sql = "SELECT id, email, first_name, last_name, age, created_at "
            + "FROM users WHERE id = ?";
        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (!rs.next()) {
                    return Optional.empty();
                }
                return Optional.of(mapUser(rs));
            }
        } catch (SQLException e) {
            throw new RuntimeException("Failed to fetch user " + id, e);
        }
    }

    public List<UserStats> getUserStats(Integer days) {
        // Normative SQL, see contracts/rest/canonical-payloads.md. The previous
        // query joined order_items, aggregated quantity*price and ordered without
        // a tiebreak, so it ran a heavier query than the other implementations and
        // its rows came back in arbitrary order among equal values.
        final String sql = """
            SELECT
                u.id AS "userId",
                u.first_name || ' ' || u.last_name AS "userName",
                COUNT(o.id) AS "totalOrders",
                COALESCE(SUM(o.total_amount), 0) AS "totalValue",
                COALESCE(AVG(o.total_amount), 0) AS "averageOrderValue"
            FROM users u
            INNER JOIN orders o ON u.id = o.user_id
                WHERE o.created_at >= NOW() - INTERVAL '1 day' * ?
            GROUP BY u.id, u.first_name, u.last_name
            ORDER BY "totalOrders" DESC, u.id
            LIMIT 100
            """;
        List<UserStats> result = new ArrayList<>();
        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, days);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    result.add(mapUserStats(rs));
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException("Failed to fetch user stats for days=" + days, e);
        }
        return result;
    }

    private static User mapUser(ResultSet rs) throws SQLException {
        User user = new User();
        user.setId(rs.getInt("id"));
        user.setEmail(rs.getString("email"));
        user.setFirstName(rs.getString("first_name"));
        user.setLastName(rs.getString("last_name"));
        user.setAge(rs.getInt("age"));
        Timestamp createdAt = rs.getTimestamp("created_at");
        user.setCreatedAt(createdAt == null ? null : createdAt.toLocalDateTime());
        return user;
    }

    private static UserStats mapUserStats(ResultSet rs) throws SQLException {
        return new UserStats(
            rs.getInt("userId"),
            rs.getString("userName"),
            rs.getInt("totalOrders"),
            rs.getDouble("totalValue"),
            rs.getDouble("averageOrderValue")
        );
    }
}

File: src/deno/fresh/src/services/database_service.ts
/**
 * Database Service - PostgreSQL Real
 * Implements connection pooling and queries to PostgreSQL
 */

import { Client } from 'https://deno.land/x/postgres@v0.19.3/mod.ts';

export interface User {
  id: number;
  email: string;
  first_name: string;
  last_name: string;
  age: number;
  created_at: Date;
}

export interface ComplexOrderResult {
  user_id: number;
  user_email: string;
  total_orders: number;
  total_value: number;
  average_value: number;
  days_since_first_order: number;
}

export interface DatabaseService {
  getUserById(id: number): Promise<User | null>;
  getComplexQuery(days: number): Promise<ComplexOrderResult[]>;
  healthCheck(): Promise<boolean>;
}

export class PostgresDatabaseService implements DatabaseService {
  private pool: any;
  private connectionString: string;

  constructor() {
    this.connectionString = Deno.env.get('DATABASE_URL') || (() => { throw new Error('DATABASE_URL is required'); })();

    // Create connection pool
    this.pool = {
      clients: [],
      max: 10,
      min: 2,
    };
  }

  async init(): Promise<void> {
    console.log('🔌 Connecting to PostgreSQL...', this.connectionString.replace(/\/\/.*@/, '//***:***@'));

    // Test connection
    const client = new Client(this.connectionString);
    await client.connect();

    // Test query
    const result = await client.query('SELECT 1 as test');
    console.log('✅ PostgreSQL connected successfully');

    await client.end();
  }

  async getUserById(id: number): Promise<User | null> {
    const client = new Client(this.connectionString);
    await client.connect();

    try {
      const result = await client.query(
        'SELECT id, email, first_name, last_name, age, created_at FROM users WHERE id = $1',
        [id]
      );

      if (result.rows.length === 0) {
        return null;
      }

      const row = result.rows[0];
      return {
        id: row.id,
        email: row.email,
        first_name: row.first_name,
        last_name: row.last_name,
        age: row.age,
        created_at: row.created_at,
      };
    } finally {
      await client.end();
    }
  }

  async getComplexQuery(days: number): Promise<ComplexOrderResult[]> {
    const client = new Client(this.connectionString);
    await client.connect();

    try {
      const query = `
        SELECT
          u.id as user_id,
          u.email as user_email,
          COUNT(DISTINCT o.id) as total_orders,
          COALESCE(SUM(oi.quantity * oi.price), 0) as total_value,
          COALESCE(AVG(oi.quantity * oi.price), 0) as average_value,
          EXTRACT(DAY FROM (NOW() - MIN(o.created_at))) as days_since_first_order
        FROM users u
        INNER JOIN orders o ON u.id = o.user_id
          AND o.created_at >= NOW() - INTERVAL '${days} days'
          AND o.status = 'completed'
        INNER JOIN order_items oi ON o.id = oi.order_id
        GROUP BY u.id, u.email
        ORDER BY total_orders DESC
        LIMIT 100
      `;

      const result = await client.query(query);

      return result.rows.map(row => ({
        user_id: row.user_id,
        user_email: row.user_email,
        total_orders: parseInt(row.total_orders),
        total_value: parseFloat(row.total_value),
        average_value: parseFloat(row.average_value),
        days_since_first_order: parseInt(row.days_since_first_order),
      }));
    } finally {
      await client.end();
    }
  }

  async healthCheck(): Promise<boolean> {
    try {
      const client = new Client(this.connectionString);
      await client.connect();
      const result = await client.query('SELECT 1');
      await client.end();
      return result.rows.length > 0;
    } catch (error) {
      console.error('❌ Database health check failed:', error);
      return false;
    }
  }
}
